### PR TITLE
Make @doc, @doc(value) and get_attribute/2, put_attribute/3 consistent

### LINKED ISSUE #4418: Odd behviour when getting and putting @doc in __on_definition__/6
### Environment

* Elixir version (elixir -v): 1.2.3
* Operating system: Ubuntu Linux 

### Current behavior

Given the code below, compiling does end up with an error message about an unmatched clause.

```exs
defmodule DocPreprocessor do
  def __on_definition__(env, _kind, _name, _args, _guards, _body) do
    mod = env.module
    doc = Module.get_attribute(mod, :doc)
    Module.put_attribute(mod, :doc, doc)
    :ok
  end
end

defmodule Example do
  @on_definition DocPreprocessor

  @doc "example"
  def foo(), do: nil
end
```

```txt
$ elixir doc_example.exs 
** (CaseClauseError) no case clause matching: [doc: "example"]
    (elixir) lib/module.ex:1085: Module.get_doc_info/2
    (elixir) lib/module.ex:973: Module.compile_doc/6
```

To make it actually work and keep the generated documentation intact you need to put `{env.line, doc}` instead.

### Expected behavior

When taking something out and putting it back in unchanged, I'd expect to work as if nothing happened at all.

It should neither be necessary to put `env.line` in the function, nor even allowed to bring the line number of the doc out of sync with the linenumber of the actual definition.

### PR DESCRIPTION
**Pull request by lexmag**
At (@) syntax works with doc value, Module functions with actual representation.

Closes #4418.

### CODE REVIEW

File: lib/elixir/lib/kernel.ex [hunk 1]
```diff
@@ -2343,7 +2344,10 @@ defmodule Kernel do
           [] -> []
           _  -> Macro.escape(stack)
         end
-        quote do: Module.get_attribute(__MODULE__, unquote(name), unquote(escaped))
+        quote do
+          attr = Module.get_attribute(__MODULE__, unquote(name), unquote(escaped))
+          __drop_line_at_doc__(unquote(name), attr)
```
Comment (NobbZ) [inline lib/elixir/lib/kernel.ex]: Return without the line. Seems to cool! (see below!)

File: lib/elixir/lib/module.ex [hunk 2]
```diff
@@ -861,6 +861,15 @@ defmodule Module do
       end
 
   """
+  def put_attribute(module, key, value)
+
+  def put_attribute(module, key, value) when key in [:moduledoc, :typedoc, :doc] do
+    unless match?({line, _} when is_integer(line), value) do
+      raise ArgumentError, "expected #{inspect(key)} attribute given in the {line, doc} format, got: #{inspect(value)}"
```
Comment (NobbZ) [inline lib/elixir/lib/module.ex]: Still requiring to put a tuple back. But making the errormessage a little bit better...

**Comment by NobbZ** [comment]
As far as I understood the code, this changes don't fix my problem at all:

> When taking something out and putting it back in unchanged, I'd expect to work as if nothing happened at all.

Thats what I have written on the mailinglist and in the report. And I still do think that way, since no said something against this expectation.

Again, if one takes the batteries out of a device and puts them back unchanged, this device crashes, but why?

**Comment by lexmag** [comment]
@NobbZ this patch actually fixes the problem described in the #4418 issue, the following code won't fail anymore:
```elixir
doc = Module.get_attribute(mod, :doc)
Module.put_attribute(mod, :doc, doc)
```
The `doc` value, which `Module.get_attribute/2` returns, is in the {line, doc} format from now on.

**Comment by NobbZ** [comment]
As far as I understand `__drop_line_at_doc__/2` it strips exactly that line information.

But now also looking at your tests (should have done that earlier), I see your point. I also do think, that I can live with that behaviour. This way I can pattern match on `Module.get_attribute/2`s return value for campatibility with older elixir as well without needing to query the version itself.

So please excuse the noise.

File: lib/elixir/lib/kernel.ex [hunk 3]
```diff
@@ -2352,6 +2356,12 @@ defmodule Kernel do
     raise ArgumentError, "expected 0 or 1 argument for @#{name}, got: #{length(args)}"
   end
 
+  @doc false
+  def __drop_line_at_doc__(:doc, {_line, doc}), do: doc
```
Comment (josevalim) [inline lib/elixir/lib/kernel.ex]: Let's please do it in the compile code and not add this function call at runtime. Call :erlang.element depending on the doc type.

File: lib/elixir/lib/module.ex (hunk 2 above)
Comment (josevalim) [inline lib/elixir/lib/module.ex]: Should we raise in preprocess attribute instead? Like here: https://github.com/lexmag/elixir/blob/master/lib/elixir/lib/module.ex#L1120?

Comment (lexmag) [inline lib/elixir/lib/module.ex]: Indeed, `preprocess_attribute` would be a better place.

**Comment by lexmag** [comment]
Updated, went with a `with` instead of `__drop_line_at_doc__ /2`.

**Pull request merged**